dcicutils/qa_utils.py

```diff
@@ -551,8 +559,150 @@ def download_fileobj(self, Bucket, Key, Fileobj, **kwargs):  # noqa - Uppercase
         Fileobj.write(data)
 
-    def download_file(self, Bucket, Key, Filename, **kwargs):
+    def download_file(self, Bucket, Key, Filename, **kwargs):  # noqa - Uppercase argument names are chosen by AWS
         if kwargs:
             raise MockKeysNotImplemented("upload_file", kwargs.keys())
         with io.open(Filename, 'wb') as fp:
             self.download_fileobj(Bucket=Bucket, Key=Key, Fileobj=fp)
+
+
+class UncustomizedInstance(Exception):
+
+    def __init__(self, instance, *, field):
+        self.instance = instance
+        self.field = field
+        declaration_class, declaration = self._find_field_declared_class(instance, field)
+        self.declaration_class = declaration_class
+        context = ""
+        if declaration_class == 'instance':
+            context = " from instance"
+        elif declaration_class:
+            context = " from class %s" % full_object_name(declaration_class)
+        message = ("Attempt to access field %s%s."
+                   " It was expected to be given a custom value in a subclass: %s."
+                   % (field, context, declaration.description))
+        super().__init__(message)
+
+    @staticmethod
+    def _find_field_declared_class(instance, field):
+        instance_value = instance.__dict__.get(field)
+        is_class = isinstance(instance, type)
+        if instance_value:
+            return instance if is_class else 'instance', instance_value
+        else:
+            for cls in instance.__mro__ if is_class else instance.__class__.__mro__:
+                cls_value = cls.__dict__.get(field)
+                if cls_value:
+                    return cls, cls_value
+            raise RuntimeError("%s does not have a field %s." % (instance, field))
+
+
+class CustomizableProperty(property):
+
+    def __init__(self, field, *, description):
+
+        self.field = field
+        self.description = description
+
+        def uncustomized(instance):
+            raise UncustomizedInstance(instance=instance, field=field)
+
+        super().__init__(uncustomized)
+
+    def __str__(self):
+        return "<%s %s>" % (self.__class__.__name__, self.field)
+
+
+def getattr_customized(thing, key):
+    # This will raise an error if the attribute is a CustomizableProperty living in the class part of the dict,
+    # but will return the object if it's in the instance.
+    value = getattr(thing, key)
+    if isinstance(value, CustomizableProperty):
+        # This is an uncustomized instance variable, not a class variable.
+        # That's not an intended use case, but just report it without involving mention of the class.
+        raise UncustomizedInstance(instance=thing, field=key)
+    else:
+        return value
+
+
+class VersionChecker:
+
+    PYPROJECT = CustomizableProperty('PYPROJECT', description="The repository-relative name of the pyproject file.")
+    CHANGELOG = CustomizableProperty('CHANGELOG', description="The repository-relative name of the change log.")
```

Comment:
It supports the ability to publish abstract classes in one library and to get decent error messages when they're used wrong in another library. I find Python does this poorly.  I might like more of our repos to use the version checker, for example, in order to promote use of changelogs. But I'm thinking it could be useful for other things that snovault might want to publish to FF and CGAP, too, so it seemed like it might be worth bundling in. It's a bit speculative to do on basis of a single use here, though I found in prior work elsewhere that once I had this facility, I used it a lot and it gave better error messages than Python presently gives if you just either use an attribute that requires complex subclass initialization across libraries or you inline the error checks every time you do. I could leave it out until we have more need if you thought that advisable. It's used in the VersionChecker class if you want to see a use.